### PR TITLE
Add GitLab CI badge support

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
Address #541

A dirty but error-prone implementation.

### CODE REVIEW

**Comment by ishitatsuyuki** [comment]
Tested. Additionally, I have implemented in a different way so that it only shows success or failed.

**Comment by bittner** [comment]
Any plans to rebase and merge this PR?

**Comment by paulmelnikow** [comment]
Hi, thanks for this contribution.

I might be missing a little context here. Is there a need for multiple auth tokens? Does a single token run into rate limits?

**Comment by ishitatsuyuki** [comment]
They haven't implement rate limiting so far.

**Comment by paulmelnikow** [comment]
In that case, you can do this a bit more simply. You can pull the secret directly from `serverSecrets`, and remove `gitlab-auth.js`. Take a look at jira or bintray as examples.

File: server.js [hunk 1]
```diff
@@ -572,6 +574,111 @@ cache(function(data, match, sendBadge, request) {
...
 }));
 
+camp.route(/^\/gitlab\/ci\/([sc])\/([^\/]+\/[^\/]+)(?:\/(.+))?\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var dataKind = match[1];
+  var userRepo = match[2];  // namespace/project
+  var branch = match[3];
+  var format = match[4];
+  var apiUrl = gitlabApiUrl + '/projects/' + encodeURIComponent(userRepo) + '/pipelines';
+  console.log(apiUrl);
```
Comment (paulmelnikow) [inline server.js]: Could you remove ^^?

File: server.js [hunk 2]
```diff
@@ -572,6 +574,111 @@ cache(function(data, match, sendBadge, request) {
...
+                sendBadge(format, badgeData);
+              });
+              break;
+            }
+            done = true;
+            break;
+          }
+        }
+        if(!done) {
+          callback(branch, page+1);
```
Comment (paulmelnikow) [inline server.js]: @espadrine has mentioned that requests need to be bounded in time and complexity.

Could this be modified to make a max of two or three requests?

File: server.js [hunk 3]
```diff
@@ -572,6 +574,111 @@ cache(function(data, match, sendBadge, request) {
...
+        }
+      } catch(e) {
+        badgeData.text[1] = 'unknown';
+        sendBadge(format, badgeData);
+      }
+    });
+  }
+  if(branch == null) {
```
Comment (paulmelnikow) [inline server.js]: To reduce complexity, could the branch default to `master` unless the user specifies a different one in the badge? Alternatively, the user could be required to specify a branch name.

File: server.js [hunk 4]
```diff
@@ -572,6 +574,111 @@ cache(function(data, match, sendBadge, request) {
...
+camp.route(/^\/gitlab\/ci\/([sc])\/([^\/]+\/[^\/]+)(?:\/(.+))?\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var dataKind = match[1];
+  var userRepo = match[2];  // namespace/project
+  var branch = match[3];
+  var format = match[4];
+  var apiUrl = gitlabApiUrl + '/projects/' + encodeURIComponent(userRepo) + '/pipelines';
+  console.log(apiUrl);
+  var badgeData = getBadgeData({'s':'build', 'c': 'coverage'}[dataKind], data);
+  var callback = function(branch, page){
```
Comment (paulmelnikow) [inline server.js]: This code is a bit difficult to follow, and may need to be reworked, it seems.

**Comment by ishitatsuyuki** [comment]
I tend to make this badge to have better behavior than the official one, since otherwise there's no point to use shields instead of their svg badge. I suspect they may implement rate limiting in the future, and copied the flexible authentication module from GitHub.

Their API sucks hard (and the documentation too), and at least two requests are required; I hate the idea of defaulting to master, since that there's many projects that use a different workflow.

I have no time to work on this for the moment, so I will update when I have free time.

**Comment by paulmelnikow** [comment]
If defaulting to master seems like a bad idea, maybe the best bet is to require the user to specify the branch.

I'll hold off until I hear from you.

**Comment by blahah** [comment]
Will this support custom gitlab installations? Would be really cool to be able to specify the root URL.

Also, see https://gitlab.com/gitlab-org/gitlab-ce/issues/30120#note_26962652

**Comment by Daniel15** [comment]
> Will this support custom gitlab installations? 

That would be good to add! Some other badges already support something similar, for example the Jenkins badge supports it (eg. https://img.shields.io/jenkins/s/https/build.dan.cx/job/yarn-e2e.svg), as does the TeamCity badge (eg. https://img.shields.io/teamcity/http/teamcity.jetbrains.com/s/bt345.svg)

**Comment by ishitatsuyuki** [comment]
I'm looking for someone to takeover the ownership of this change: simply comment here, and I will transfer my fork to you. I have no time to work on this.

**Comment by paulmelnikow** [comment]
Thanks @ishitatsuyuki.

Since this needs significant work, another option is to link to it from #541 and close it out. The code will be here, attached to the PR, whenever someone wants to pick it up.

**Comment by paulmelnikow** [comment]
@blahah Would you like to pick this up? There's an issue which needs to be solved, I think with gitlab, so that this can make a finite, small number of API calls.

**Comment by blahah** [comment]
@paulmelnikow happy to - it will be a few days before I can commit time to it, but after that I'll push it over the finish line. Thanks to everyone who's worked on this so far.

**Comment by ishitatsuyuki** [comment]
@blahah I have requested for a transfer to you. Hope you can have luck doing this.

I will still be subscribed for a while, so comment in case of problems.

**Comment by paulmelnikow** [comment]
Wanted to mention we have a new service testing framework, so it would be good if the new version of this PR included tests: https://github.com/badges/shields/tree/master/service-tests

**Comment by paulmelnikow** [comment]
@blahah would you have some time to address these comments and add tests? Please feel free to grab the branch, add more commits addressing the comments, and open a new PR referencing this one. If anyone else would like to adopt it, please feel free to do the same. Will close in one month if no one has.

**Comment by paulmelnikow** [comment]
Closing for inactivity. If you're reading this and want this implemented, don't be discouraged. You can grab the commits from the pull request branch using `git fetch origin pull/841/head:pr-841`, address the comments, and reopen.

**Pull request closed without merging**